span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/RpcFieldsGenerator.java

```diff
@@ -0,0 +1,125 @@
+  }
+
+  private static Map<String, FieldGenerator<Rpc.Builder>> initializeFieldGenerators() {
+    Map<String, FieldGenerator<Rpc.Builder>> fieldGeneratorMap = new HashMap<>();
+    fieldGeneratorMap.put(
+        OTEL_SPAN_TAG_RPC_SYSTEM_ATTR,
+        (key, keyValue, builder, tagsMap) -> {
+          builder.setSystem(ValueConverter.getString(keyValue));
+        }
+    );
+    fieldGeneratorMap.put(
+        RawSpanConstants.getValue(OTEL_SPAN_TAG_RPC_SERVICE),
+        (key, keyValue, builder, tagsMap) -> {
+          builder.setService(ValueConverter.getString(keyValue));
+        }
+    );
+    fieldGeneratorMap.put(
+        RawSpanConstants.getValue(OTEL_SPAN_TAG_RPC_METHOD),
+        (key, keyValue, builder, tagsMap) -> {
+          builder.setMethod(ValueConverter.getString(keyValue));
+        }
+    );
+
+    return fieldGeneratorMap;
+  }
+
+  @Override
+  protected Rpc.Builder getProtocolBuilder(Event.Builder eventBuilder) {
+    return eventBuilder.getRpcBuilder();
+  }
+
+  @Override
+  protected Map<String, FieldGenerator<Rpc.Builder>> getFieldGeneratorMap() {
+    return fieldGeneratorMap;
+  }
+
+  public boolean handleKeyIfNecessary(String key, JaegerSpanInternalModel.KeyValue keyValue,
+                                      Event.Builder eventBuilder,
+                                      Map<String, JaegerSpanInternalModel.KeyValue> tagsMap) {
+    String rpcSystem = getRpcSystem(tagsMap);
+    if (!isRpcSystemGrpc(rpcSystem)) {
+      return false;
+    }
+    Grpc.Builder grpcProtocolBuilder = grpcFieldsGenerator.getProtocolBuilder(eventBuilder);
+    Map<String, FieldGenerator<Grpc.Builder>> rpcFieldGeneratorMap = grpcFieldsGenerator.getRpcFieldGeneratorMap();
+    FieldGenerator<Grpc.Builder> rpcFieldGenerator = rpcFieldGeneratorMap.get(key);
+    if (rpcFieldGenerator != null) {
+      rpcFieldGenerator.run(key, keyValue, grpcProtocolBuilder, tagsMap);
+      return true;
+    }
+    if (key.startsWith(REQUEST_METADATA_PREFIX)) {
+      getSuffix(key, REQUEST_METADATA_PREFIX).ifPresent(
+          metadata -> grpcFieldsGenerator.handleRpcRequestMetadata(metadata, keyValue, grpcProtocolBuilder)
+      );
+      return true;
+    } else if (key.startsWith(RESPONSE_METADATA_PREFIX)) {
+      getSuffix(key, RESPONSE_METADATA_PREFIX).ifPresent(
+          metadata -> grpcFieldsGenerator.handleRpcResponseMetadata(metadata, keyValue, grpcProtocolBuilder)
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private String getRpcSystem(Map<String, JaegerSpanInternalModel.KeyValue> tagsMap) {
+    JaegerSpanInternalModel.KeyValue keyValue = tagsMap.get(OTEL_SPAN_TAG_RPC_SYSTEM_ATTR);
+    if (keyValue != null) {
+      String val = ValueConverter.getString(keyValue);
+      return val;
+    }
+    return null;
+  }
+
+  private boolean isRpcSystemGrpc(String rpcSystem) {
+    if (StringUtils.isNotBlank(rpcSystem) && StringUtils.equals(rpcSystem, OTEL_RPC_SYSTEM_GRPC_VAL)) {
```

Comment:
Return with condition directly instead of `if` statement?

Reply:
Done